File: 400-600/416. Partition Equal Subset Sum/solution.cpp
//solution 1
// n is the length of nums, w is the half of sum
//standard O(n*w)
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum = accumulate(nums.begin(),nums.end(),0);
        if (sum%2) return false;
        int target = sum/2;
        vector<bool> dp(target+1,false);
        dp[0] = true;
        
        for (auto val: nums) {
            for (int j=target-1; j>=0; j--) {
                if(j+val<=target) dp[j+val] = dp[j+val] || dp[j];
            }
        }
        return dp.back();
    }
};

//solution 2
//time: O(n*(w/32))
constexpr int x = 100*100 + 1;

class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum = accumulate(nums.begin(),nums.end(),0);
        if (sum%2) return false;
        bitset<x> dp(1);
        
        for (auto val: nums) {
            dp |= dp << val;
        }
        return dp[sum/2];
    }
};

File: 400-600/416. Partition Equal Subset Sum/solution.js
//recursion
//time: O(len*half)
//space: O(len*half)
var canPartition = function(nums) {
    let sum = 0
    for (let num of nums) {
        sum += num
    }
    if (sum%2) return false
    return helper(nums,sum/2)
};

let helper = (nums,half,index=0,memo={}) => {
    if (half in memo && index in memo[half]) return memo[half][index]
    if (!memo[half]) memo[half] = {}
    if (half<=0 || index === nums.length) {
        memo[half][index] = (half === 0)
        return memo[half][index]
    } else {
        memo[half][index] = helper(nums,half,index+1,memo) || helper(nums,half-nums[index],index+1,memo)
        return memo[half][index]
    }
}



//tabulation
//time: O(len*half)
//space: O(half)
var canPartition2 = function(nums) {
    let sum = 0
    for (let num of nums) {
        sum += num
    }
    if (sum%2) return false
    let half = sum/2
    let dp = new Array(half+1).fill(false)
    dp[0] = true
    for (let num of nums) {
        for (let i=half; i>=0; i--) {
            if (i >= num) dp[i] = dp[i] || dp[i-num] 
        }
    }
    return dp[half]
};

let nums = [1,5,11,5]
console.log(canPartition2(nums))
